docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,156 @@
+
+- Parallelized initialization of components.
+
+## Questions
+
+- To what extent does this new initialization framework need to accommodate
+  the legacy p2p stack? Would it be possible to delay a great deal of this
+  work to the 0.36 cycle to avoid this complexity? 
+  
+  - Answer: _depends on timing_, and the requirement to ship pluggable
+    reactors in 0.35. 
+  
+- There's clear and common interest in injecting some kind of mempool, and
+  intuitively this makes sense as the highest priority.
+  
+  - Does the existing `mempool.Mempool` interface make sense to expose and
+    commit to, or should we make a smaller interface? 
+    
+	- Answer: _yes_, smaller mempool interface is better. 
+
+  - Is it reasonable for users to bring their own mempool without bringing
+    their own reactor?
+	
+	- Answer: _probably not_, but this might be a function of how the
+      current implementations work and can be refactored. 
+
+- Are there reactors, other than the mempool, that make sense to support
+  ad-hoc replacing (e.g. statesync, blockchain (fastsync),) and if so should
+  we provide any first-class support for this?
+
+  - Answer: _defer_. Certainly this shouldn't be attempted until after
+    there is a refactor of the consensus reactor. Is more likely than
+    block/fast sync. 
+
+- There's a dependency (and nearly a cycle), between consensus,
+  blockchain (fastsync), which makes it very hard to allow meaningful
+  injection of any of these components without allowing injection of
+  all of them (statesync provider as a possible exception.) Is this
+  acceptable?
+  
+  - Answer: _yes_. 
+
+## Future Work
+
+- Improve or simplify the `service.Service` interface. There are some
+  pretty clear limitations with this interface as written (there's no
+  way to timeout slow startup or shut down, the cycle between the
+  `service.BaseService` and `service.Service` implementations is
+  troubling, the default panic in `OnReset` seems troubling.)
+
+- Support explicit dependencies between components and allow for
+  parallel startup, so that different reactors can startup at the same
+  time, where possible.
+
+## Proposal / Conclusion 
+
+- Wait for e2e tests to stablize, before changing node initialization.
+
+- Continue building on [the experimental branch
+  branch](https://github.com/tendermint/tendermint/tree/tychoish/scratch-node-minimize)
+  to reduce the complexity in the node initialization, and recast
+  `nodeImpl` as a list of `service.Service` objects, modulo
+  requirements for legacy p2p components which will be deleted.
+
+- Add the ability to construct a tendermint node via a new public
+  constructor that will take a list of additional services that will
+  either be appended to the end of the node initialization process
+  (simple,) or (based on name) replace one of the default services
+  (MVP for mempool only.)
+
+- Move components from `p2p` into `types` to support writing
```

Comment:
Not a fan of this. We should not defer to types. 

Nit: can we follow the adr template provided in the repo?

Reply:
🙏  just move things to `pkg/<domain>` please!

Reply:
I don't think that removing `types` in favor of `pkg` should be a prerequisite to this project or effort, nor do I think that it's reasonable to expand the scope of the project to cover that work.

I agree that types isn't a good pattern to use, and that we should move away from it, but I'm not sure that it makes sense to tie these projects together given the release schedule/goals. Let me think about it a bit.

Reply:
I called this out as an open question, and think we should decide on that separately from this ADR, though I think it impacts a few of the later-stage questions.